protocol/contracts/v2/TreasuryPriceIndexOracle.sol

```diff
@@ -21,98 +20,148 @@ contract TreasuryPriceIndexOracle is ITreasuryPriceIndexOracle, TempleElevatedAc
+        maxAbsTreasuryPriceIndexRateOfChange = _maxAbsTreasuryPriceIndexRateOfChange;
     }
 
     /**
      * @notice The current Treasury Price Index (TPI) value
-     * @dev If the TPI has just been updated, the old TPI will be used until `cooldownSecs` has elapsed
      */
     function treasuryPriceIndex() public override view returns (uint96) {
-        return (block.timestamp < (tpiData.lastUpdatedAt + tpiData.cooldownSecs))
-            ? tpiData.previousTpi  // use the previous TPI if we haven't passed the cooldown yet.
-            : tpiData.currentTpi;  // use the new TPI
+        uint32 _now = uint32(block.timestamp);
+        if (_now >= tpiData.targetTime) {
+            // Target date reached, no calculation required just return the target TPI
+            return tpiData.targetTpi;  
+        } else {
+            unchecked {
+                int96 delta = tpiData.tpiSlope * int32(_now - tpiData.startTime);
+                return uint96(delta + int96(tpiData.startingTpi));
+            }
+        }
+     }
+
+    /**
+     * @notice Set the maximum allowed TPI change on any single `setTreasuryPriceIndexAt()`, in absolute terms
+     * between the TPI as of now and the targetTpi
+     * @dev 18 decimal places, 0.20e18 == $0.20
+     */
+    function setMaxTreasuryPriceIndexDelta(uint96 maxDelta) external override onlyElevatedAccess {
+        emit MaxTreasuryPriceIndexDeltaSet(maxDelta);
+        maxTreasuryPriceIndexDelta = maxDelta;
     }
 
     /**
-     * @notice Set the number of seconds to elapse before a new TPI will take effect.
+     * @notice Set the minimum time delta required for TPI to reach it's target value when 
+     * `setTreasuryPriceIndexAt()` is called.
+     * @dev In seconds.
      */
-    function setTpiCooldown(uint32 cooldownSecs) external override onlyElevatedAccess {
-        emit TpiCooldownSet(cooldownSecs);
-        tpiData.cooldownSecs = cooldownSecs;
+    function setMinTreasuryPriceIndexTargetTimeDelta(uint32 minTargetTimeDelta) external override onlyElevatedAccess {
+        emit MinTreasuryPriceIndexTargetTimeDeltaSet(minTargetTimeDelta);
+        minTreasuryPriceIndexTargetTimeDelta = minTargetTimeDelta;
     }
 
     /**
-     * @notice Set the maximum allowed TPI change on any single `setTreasuryPriceIndex()`, in absolute terms.
-     * @dev 18 decimal places, 0.20e18 == $0.20
+     * @notice Set the maximum absolute rate of change of TPI allowed, when 
+     * `setTreasuryPriceIndexAt()` is called.
+     * @dev Units: [TPI / second]
      */
-    function setMaxTreasuryPriceIndexDelta(uint256 maxDelta) external override onlyElevatedAccess {
-        emit MaxTreasuryPriceIndexDeltaSet(maxDelta);
-        maxTreasuryPriceIndexDelta = maxDelta;
+    function setMaxAbsTreasuryPriceIndexRateOfChange(uint96 tpiDelta, uint32 timeDelta) external override onlyElevatedAccess {
+        // Calculate the rate of change, rounding down.
+        uint96 maxAbsRateOfChange = tpiDelta / timeDelta;
```

Comment:
I personally see it as a bit over the top - it just means the next time `setTreasuryPriceIndexAt()` is called it will fail. This does sort of give an extra mechanism to pause any further increases which could be useful.

Reply:
> I personally see it as a bit over the top - it just means the next time `setTreasuryPriceIndexAt()` is called it will fail. This does sort of give an extra mechanism to pause any further increases which could be useful.

Good point